File: app/src/main/java/com/meituy/app/FilterEngine.kt
package com.meituy.app

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import kotlin.math.sqrt

object FilterEngine {

    fun applyFilter(bitmap: Bitmap, filterType: FilterType, intensity: Float = 1.0f): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        val result = bitmap.copy(Bitmap.Config.ARGB_8888, true) ?: return bitmap

        when (filterType) {
            FilterType.ORIGINAL -> return result
            FilterType.RICON_FLASH -> applyRiconFlash(result, width, height, intensity)
            FilterType.FLASH_FILM -> applyFlashFilm(result, width, height, intensity)
            FilterType.G7X -> applyG7X(result, width, height, intensity)
            FilterType.FUJI_FLASH -> applyFujiFlash(result, width, height, intensity)
            FilterType.GOLDEN_HOUR -> applyGoldenHour(result, width, height, intensity)
            FilterType.MATAHARI_TERBENAM -> applyMatahariTerbenam(result, width, height, intensity)
            FilterType.LAMPU_KILAT_IPHONE -> applyLampuKilatIphone(result, width, height, intensity)
        }

        return result
    }

    // Interpolasi dari identitas: intensity 0 = gambar asli, 1 = kekuatan penuh.
    private fun blendedMatrix(gains: FloatArray, offsets: FloatArray, intensity: Float): ColorMatrix {
        return ColorMatrix(floatArrayOf(
            1f + (gains[0] - 1f) * intensity, 0.0f, 0.0f, 0f, offsets[0] * intensity,
            0.0f, 1f + (gains[1] - 1f) * intensity, 0.0f, 0f, offsets[1] * intensity,
            0.0f, 0.0f, 1f + (gains[2] - 1f) * intensity, 0f, offsets[2] * intensity,
            0.0f, 0.0f, 0.0f, 1f, 0f
        ))
    }

    private fun drawWithMatrix(bitmap: Bitmap, matrix: ColorMatrix) {
        val paint = Paint().apply { colorFilter = ColorMatrixColorFilter(matrix) }
        Canvas(bitmap).drawBitmap(bitmap, 0f, 0f, paint)
    }

    private fun applyRiconFlash(bitmap: Bitmap, width: Int, height: Int, intensity: Float) {
        drawWithMatrix(bitmap, blendedMatrix(
            gains = floatArrayOf(1.2f, 1.15f, 1.0f),
            offsets = floatArrayOf(20f, 15f, 0f),
            intensity = intensity
        ))

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        for (i in pixels.indices) {
            val pixel = pixels[i]
            var r = Color.red(pixel)
            var g = Color.green(pixel)
            var b = Color.blue(pixel)

            val lum = 0.299 * r + 0.587 * g + 0.114 * b

            if (lum > 180) {
                r = (r + (255 - r) * 0.15f * intensity).toInt().coerceIn(0, 255)
                g = (g + (255 - g) * 0.12f * intensity).toInt().coerceIn(0, 255)
            }

            if (lum < 60) {
                val lift = (60 - lum) * 0.08f * intensity
                r = (r + lift).toInt().coerceIn(0, 255)
                g = (g + lift * 0.9f).toInt().coerceIn(0, 255)
            }

            pixels[i] = Color.argb(Color.alpha(pixel), r, g, b)
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
    }

    private fun applyFlashFilm(bitmap: Bitmap, width: Int, height: Int, intensity: Float) {
        drawWithMatrix(bitmap, blendedMatrix(
            gains = floatArrayOf(1.2f, 1.08f, 1.05f),
            offsets = floatArrayOf(10f, 5f, 0f),
            intensity = intensity
        ))

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        for (i in pixels.indices) {
            val pixel = pixels[i]
            var r = Color.red(pixel)
            var g = Color.green(pixel)
            var b = Color.blue(pixel)

            val lum = 0.299 * r + 0.587 * g + 0.114 * b

            val contrast = 1.0f + 0.15f * intensity
            r = (((r / 255.0 - 0.5) * contrast + 0.5) * 255).toInt().coerceIn(0, 255)
            g = (((g / 255.0 - 0.5) * contrast + 0.5) * 255).toInt().coerceIn(0, 255)
            b = (((b / 255.0 - 0.5) * contrast + 0.5) * 255).toInt().coerceIn(0, 255)

            if (lum > 200) {
                r = (r + (255 - r) * 0.1f * intensity).toInt().coerceIn(0, 255)
                g = (g + (255 - g) * 0.08f * intensity).toInt().coerceIn(0, 255)
            }

            pixels[i] = Color.argb(Color.alpha(pixel), r, g, b)
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
    }

    private fun applyG7X(bitmap: Bitmap, width: Int, height: Int, intensity: Float) {
        drawWithMatrix(bitmap, blendedMatrix(
            gains = floatArrayOf(1.25f, 1.2f, 1.17f),
            offsets = floatArrayOf(12f, 8f, 10f),
            intensity = intensity
        ))

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        for (i in pixels.indices) {
            val pixel = pixels[i]
            var r = Color.red(pixel)
            var g = Color.green(pixel)
            var b = Color.blue(pixel)

            val lum = 0.299 * r + 0.587 * g + 0.114 * b

            if (lum > 190) {
                val roll = (lum - 190) / 65.0
                r = (r + (255 - r) * roll * 0.12f * intensity).toInt().coerceIn(0, 255)
                g = (g + (255 - g) * roll * 0.10f * intensity).toInt().coerceIn(0, 255)
                b = (b + (255 - b) * roll * 0.08f * intensity).toInt().coerceIn(0, 255)
            }

            if (lum < 50) {
                val lift = (50 - lum) * 0.06f * intensity
                r = (r + lift).toInt().coerceIn(0, 255)
                g = (g + lift * 1.1f).toInt().coerceIn(0, 255)
                b = (b + lift * 1.2f).toInt().coerceIn(0, 255)
            }

            pixels[i] = Color.argb(Color.alpha(pixel), r, g, b)
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
    }

    private fun applyFujiFlash(bitmap: Bitmap, width: Int, height: Int, intensity: Float) {
        drawWithMatrix(bitmap, blendedMatrix(
            gains = floatArrayOf(1.12f, 1.18f, 1.08f),
            offsets = floatArrayOf(8f, 12f, 5f),
            intensity = intensity
        ))

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        for (i in pixels.indices) {
            val pixel = pixels[i]
            var r = Color.red(pixel)
            var g = Color.green(pixel)
            var b = Color.blue(pixel)

            val lum = 0.299 * r + 0.587 * g + 0.114 * b

            if (lum > 180) {
                r = (r + (255 - r) * 0.08f * intensity).toInt().coerceIn(0, 255)
                g = (g + (255 - g) * 0.12f * intensity).toInt().coerceIn(0, 255)
            }

            if (lum < 55) {
                val lift = (55 - lum) * 0.07f * intensity
                r = (r + lift * 1.1f).toInt().coerceIn(0, 255)
                g = (g + lift).toInt().coerceIn(0, 255)
                b = (b + lift * 0.9f).toInt().coerceIn(0, 255)
            }

            pixels[i] = Color.argb(Color.alpha(pixel), r, g, b)
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
    }

    private fun applyGoldenHour(bitmap: Bitmap, width: Int, height: Int, intensity: Float) {
        drawWithMatrix(bitmap, blendedMatrix(
            gains = floatArrayOf(1.35f, 1.15f, 0.9f),
            offsets = floatArrayOf(25f, 15f, -10f),
            intensity = intensity
        ))

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        for (i in pixels.indices) {
            val pixel = pixels[i]
            var r = Color.red(pixel)
            var g = Color.green(pixel)
            var b = Color.blue(pixel)

            val lum = 0.299 * r + 0.587 * g + 0.114 * b

            if (lum > 150) {
                val warmth = (lum - 150) / 105.0
                r = (r + warmth * 15 * intensity).toInt().coerceIn(0, 255)
                g = (g + warmth * 8 * intensity).toInt().coerceIn(0, 255)
                b = (b - warmth * 10 * intensity).toInt().coerceIn(0, 255)
            }

            if (lum < 70) {
                val lift = (70 - lum) * 0.06f * intensity
                r = (r + lift * 1.2f).toInt().coerceIn(0, 255)
                g = (g + lift * 0.9f).toInt().coerceIn(0, 255)
                b = (b + lift * 0.7f).toInt().coerceIn(0, 255)
            }

            pixels[i] = Color.argb(Color.alpha(pixel), r, g, b)
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
    }

    private fun applyMatahariTerbenam(bitmap: Bitmap, width: Int, height: Int, intensity: Float) {
        drawWithMatrix(bitmap, blendedMatrix(
            gains = floatArrayOf(1.45f, 1.05f, 0.88f),
            offsets = floatArrayOf(30f, 10f, -15f),
            intensity = intensity
        ))

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        for (i in pixels.indices) {
            val pixel = pixels[i]
            var r = Color.red(pixel)
            var g = Color.green(pixel)
            var b = Color.blue(pixel)

            val lum = 0.299 * r + 0.587 * g + 0.114 * b

            if (lum > 130) {
                val warmth = (lum - 130) / 125.0
                r = (r + warmth * 20 * intensity).toInt().coerceIn(0, 255)
                g = (g + warmth * 10 * intensity).toInt().coerceIn(0, 255)
                b = (b - warmth * 15 * intensity).toInt().coerceIn(0, 255)
            }

            if (lum < 60) {
                val lift = (60 - lum) * 0.05f * intensity
                r = (r + lift * 1.3f).toInt().coerceIn(0, 255)
                g = (g + lift * 0.8f).toInt().coerceIn(0, 255)
                b = (b + lift * 0.6f).toInt().coerceIn(0, 255)
            }

            pixels[i] = Color.argb(Color.alpha(pixel), r, g, b)
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)

        if (intensity > 0.3f) {
            applyVignette(bitmap, width, height, 0.25f * intensity)
        }
    }

    private fun applyLampuKilatIphone(bitmap: Bitmap, width: Int, height: Int, intensity: Float) {
        drawWithMatrix(bitmap, blendedMatrix(
            gains = floatArrayOf(1.18f, 1.16f, 1.15f),
            offsets = floatArrayOf(15f, 12f, 8f),
            intensity = intensity
        ))

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        for (i in pixels.indices) {
            val pixel = pixels[i]
            var r = Color.red(pixel)
            var g = Color.green(pixel)
            var b = Color.blue(pixel)

            val lum = 0.299 * r + 0.587 * g + 0.114 * b

            if (lum > 200) {
                r = (r + (255 - r) * 0.1f * intensity).toInt().coerceIn(0, 255)
                g = (g + (255 - g) * 0.09f * intensity).toInt().coerceIn(0, 255)
                b = (b + (255 - b) * 0.08f * intensity).toInt().coerceIn(0, 255)
            }

            if (lum < 40) {
                val lift = (40 - lum) * 0.07f * intensity
                r = (r + lift).toInt().coerceIn(0, 255)
                g = (g + lift).toInt().coerceIn(0, 255)
                b = (b + lift * 1.1f).toInt().coerceIn(0, 255)
            }

            pixels[i] = Color.argb(Color.alpha(pixel), r, g, b)
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
    }

    private fun applyVignette(bitmap: Bitmap, width: Int, height: Int, strength: Float) {
        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        val centerX = width / 2f
        val centerY = height / 2f

        for (y in 0 until height) {
            for (x in 0 until width) {
                val idx = y * width + x
                val pixel = pixels[idx]

                val dx = (x - centerX) / centerX
                val dy = (y - centerY) / centerY
                val dist = sqrt(dx * dx + dy * dy)

                val vignette = 1.0f - (dist * strength).coerceIn(0f, 1f)

                val r = (Color.red(pixel) * vignette).toInt().coerceIn(0, 255)
                val g = (Color.green(pixel) * vignette).toInt().coerceIn(0, 255)
                val b = (Color.blue(pixel) * vignette).toInt().coerceIn(0, 255)

                pixels[idx] = Color.argb(Color.alpha(pixel), r, g, b)
            }
        }

        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
    }
}

enum class FilterType(val displayName: String) {
    ORIGINAL("Original"),
    RICON_FLASH("Ricon Flash"),
    FLASH_FILM("Flash Film"),
    G7X("G7X"),
    FUJI_FLASH("Fuji Flash"),
    GOLDEN_HOUR("Golden Hour"),
    MATAHARI_TERBENAM("Matahari Terbenam"),
    LAMPU_KILAT_IPHONE("Lampu Kilat iPhone")
}

File: lib/engine/filter_engine.dart
import 'dart:math' as math;
import 'dart:typed_data';

import '../models/filter_type.dart';

class FilteredImage {
  final Uint8List rgba;
  final int width;
  final int height;

  const FilteredImage(this.rgba, this.width, this.height);
}

/// Port 1:1 dari FilterEngine.kt: color-matrix blend dari identitas,
/// lalu threshold highlight/shadow per piksel. Urutan operasi dipertahankan.
class FilterEngine {
  const FilterEngine._();

  static FilteredImage apply(
    Uint8List rgba,
    int width,
    int height,
    FilterType type,
    double intensity,
  ) {
    if (type == FilterType.original || intensity <= 0) {
      return FilteredImage(Uint8List.fromList(rgba), width, height);
    }
    final out = Uint8List.fromList(rgba);
    switch (type) {
      case FilterType.original:
        break;
      case FilterType.riconFlash:
        _riconFlash(out, intensity);
        break;
      case FilterType.flashFilm:
        _flashFilm(out, intensity);
        break;
      case FilterType.g7x:
        _g7x(out, intensity);
        break;
      case FilterType.fujiFlash:
        _fujiFlash(out, intensity);
        break;
      case FilterType.goldenHour:
        _goldenHour(out, intensity);
        break;
      case FilterType.matahariTerbenam:
        _matahariTerbenam(out, width, height, intensity);
        break;
      case FilterType.lampuKilatIphone:
        _lampuKilatIphone(out, intensity);
        break;
    }
    return FilteredImage(out, width, height);
  }

  /// Matrix blend dari identitas: intensity 0 = gambar asli, 1 = penuh.
  static void _blendedMatrixPass(
    Uint8List px,
    List<double> gains,
    List<double> offsets,
    double t,
  ) {
    final gr = 1.0 + (gains[0] - 1.0) * t;
    final gg = 1.0 + (gains[1] - 1.0) * t;
    final gb = 1.0 + (gains[2] - 1.0) * t;
    final or_ = offsets[0] * t;
    final og = offsets[1] * t;
    final ob = offsets[2] * t;
    for (var i = 0; i < px.length; i += 4) {
      px[i] = (px[i] * gr + or_).round().clamp(0, 255);
      px[i + 1] = (px[i + 1] * gg + og).round().clamp(0, 255);
      px[i + 2] = (px[i + 2] * gb + ob).round().clamp(0, 255);
    }
  }

  static void _riconFlash(Uint8List px, double t) {
    _blendedMatrixPass(px, [1.2, 1.15, 1.0], [20, 15, 0], t);
    for (var i = 0; i < px.length; i += 4) {
      final r = px[i].toDouble(), g = px[i + 1].toDouble(), b = px[i + 2].toDouble();
      final lum = 0.299 * r + 0.587 * g + 0.114 * b;
      var nr = r, ng = g;
      if (lum > 180) {
        nr = r + (255 - r) * 0.15 * t;
        ng = g + (255 - g) * 0.12 * t;
      }
      if (lum < 60) {
        final lift = (60 - lum) * 0.08 * t;
        nr += lift;
        ng += lift * 0.9;
      }
      px[i] = nr.round().clamp(0, 255);
      px[i + 1] = ng.round().clamp(0, 255);
    }
  }

  static void _flashFilm(Uint8List px, double t) {
    _blendedMatrixPass(px, [1.2, 1.08, 1.05], [10, 5, 0], t);
    final contrast = 1.0 + 0.15 * t;
    for (var i = 0; i < px.length; i += 4) {
      final r = px[i].toDouble(), g = px[i + 1].toDouble(), b = px[i + 2].toDouble();
      final lum = 0.299 * r + 0.587 * g + 0.114 * b;
      var nr = (((r / 255.0 - 0.5) * contrast + 0.5) * 255);
      var ng = (((g / 255.0 - 0.5) * contrast + 0.5) * 255);
      var nb = (((b / 255.0 - 0.5) * contrast + 0.5) * 255);
      if (lum > 200) {
        nr += (255 - nr) * 0.1 * t;
        ng += (255 - ng) * 0.08 * t;
      }
      px[i] = nr.round().clamp(0, 255);
      px[i + 1] = ng.round().clamp(0, 255);
      px[i + 2] = nb.round().clamp(0, 255);
    }
  }

  static void _g7x(Uint8List px, double t) {
    _blendedMatrixPass(px, [1.25, 1.2, 1.17], [12, 8, 10], t);
    for (var i = 0; i < px.length; i += 4) {
      final r = px[i].toDouble(), g = px[i + 1].toDouble(), b = px[i + 2].toDouble();
      final lum = 0.299 * r + 0.587 * g + 0.114 * b;
      var nr = r, ng = g, nb = b;
      if (lum > 190) {
        final roll = (lum - 190) / 65.0;
        nr += (255 - r) * roll * 0.12 * t;
        ng += (255 - g) * roll * 0.10 * t;
        nb += (255 - b) * roll * 0.08 * t;
      }
      if (lum < 50) {
        final lift = (50 - lum) * 0.06 * t;
        nr += lift;
        ng += lift * 1.1;
        nb += lift * 1.2;
      }
      px[i] = nr.round().clamp(0, 255);
      px[i + 1] = ng.round().clamp(0, 255);
      px[i + 2] = nb.round().clamp(0, 255);
    }
  }

  static void _fujiFlash(Uint8List px, double t) {
    _blendedMatrixPass(px, [1.12, 1.18, 1.08], [8, 12, 5], t);
    for (var i = 0; i < px.length; i += 4) {
      final r = px[i].toDouble(), g = px[i + 1].toDouble(), b = px[i + 2].toDouble();
      final lum = 0.299 * r + 0.587 * g + 0.114 * b;
      var nr = r, ng = g, nb = b;
      if (lum > 180) {
        nr += (255 - r) * 0.08 * t;
        ng += (255 - g) * 0.12 * t;
      }
      if (lum < 55) {
        final lift = (55 - lum) * 0.07 * t;
        nr += lift * 1.1;
        ng += lift;
        nb += lift * 0.9;
      }
      px[i] = nr.round().clamp(0, 255);
      px[i + 1] = ng.round().clamp(0, 255);
      px[i + 2] = nb.round().clamp(0, 255);
    }
  }

  static void _goldenHour(Uint8List px, double t) {
    _blendedMatrixPass(px, [1.35, 1.15, 0.9], [25, 15, -10], t);
    for (var i = 0; i < px.length; i += 4) {
      final r = px[i].toDouble(), g = px[i + 1].toDouble(), b = px[i + 2].toDouble();
      final lum = 0.299 * r + 0.587 * g + 0.114 * b;
      var nr = r, ng = g, nb = b;
      if (lum > 150) {
        final warmth = (lum - 150) / 105.0;
        nr += warmth * 15 * t;
        ng += warmth * 8 * t;
        nb -= warmth * 10 * t;
      }
      if (lum < 70) {
        final lift = (70 - lum) * 0.06 * t;
        nr += lift * 1.2;
        ng += lift * 0.9;
        nb += lift * 0.7;
      }
      px[i] = nr.round().clamp(0, 255);
      px[i + 1] = ng.round().clamp(0, 255);
      px[i + 2] = nb.round().clamp(0, 255);
    }
  }

  static void _matahariTerbenam(Uint8List px, int width, int height, double t) {
    _blendedMatrixPass(px, [1.45, 1.05, 0.88], [30, 10, -15], t);
    for (var i = 0; i < px.length; i += 4) {
      final r = px[i].toDouble(), g = px[i + 1].toDouble(), b = px[i + 2].toDouble();
      final lum = 0.299 * r + 0.587 * g + 0.114 * b;
      var nr = r, ng = g, nb = b;
      if (lum > 130) {
        final warmth = (lum - 130) / 125.0;
        nr += warmth * 20 * t;
        ng += warmth * 10 * t;
        nb -= warmth * 15 * t;
      }
      if (lum < 60) {
        final lift = (60 - lum) * 0.05 * t;
        nr += lift * 1.3;
        ng += lift * 0.8;
        nb += lift * 0.6;
      }
      px[i] = nr.round().clamp(0, 255);
      px[i + 1] = ng.round().clamp(0, 255);
      px[i + 2] = nb.round().clamp(0, 255);
    }
    if (t > 0.3) _vignette(px, width, height, 0.25 * t);
  }

  static void _lampuKilatIphone(Uint8List px, double t) {
    _blendedMatrixPass(px, [1.18, 1.16, 1.15], [15, 12, 8], t);
    for (var i = 0; i < px.length; i += 4) {
      final r = px[i].toDouble(), g = px[i + 1].toDouble(), b = px[i + 2].toDouble();
      final lum = 0.299 * r + 0.587 * g + 0.114 * b;
      var nr = r, ng = g, nb = b;
      if (lum > 200) {
        nr += (255 - r) * 0.1 * t;
        ng += (255 - g) * 0.09 * t;
        nb += (255 - b) * 0.08 * t;
      }
      if (lum < 40) {
        final lift = (40 - lum) * 0.07 * t;
        nr += lift;
        ng += lift;
        nb += lift * 1.1;
      }
      px[i] = nr.round().clamp(0, 255);
      px[i + 1] = ng.round().clamp(0, 255);
      px[i + 2] = nb.round().clamp(0, 255);
    }
  }

  static void _vignette(Uint8List px, int width, int height, double strength) {
    final centerX = width / 2.0;
    final centerY = height / 2.0;
    for (var y = 0; y < height; y++) {
      final dy = (y - centerY) / centerY;
      for (var x = 0; x < width; x++) {
        final dx = (x - centerX) / centerX;
        final dist = math.sqrt(dx * dx + dy * dy);
        final vig = 1.0 - (dist * strength).clamp(0.0, 1.0);
        final idx = (y * width + x) * 4;
        px[idx] = (px[idx] * vig).round().clamp(0, 255);
        px[idx + 1] = (px[idx + 1] * vig).round().clamp(0, 255);
        px[idx + 2] = (px[idx + 2] * vig).round().clamp(0, 255);
      }
    }
  }
}
